subsys/usb/device_next/class/usbd_cdc_acm.c

```diff
@@ -114,6 +115,8 @@ struct cdc_acm_uart_data {
 	struct k_work rx_fifo_work;
 	atomic_t state;
 	struct k_sem notif_sem;
+	/* USBD CDC ACM TX transaction deferred due to BULK IN EP busy */
+	bool tx_deferred;
```

Comment:
Why new variable instead of just using `data->tx_fifo.altered` like it was earlier?

Reply:
Hi @tmon-nordic, 

I was just about to add my report on what I have done here. 

Since that #75439 has been merged into zephyr main, I have rebased my works and retested with the changes of #75439 applied. However, I am still able to reproduce this issue. 

I have also run more stress test on my fixed ported over to the latest main, and found out that raising `data->tx_fifo.altered` wouldn't help in retrigger the TX transaction, as `tx_fifo.altered` is cleared in `cdc_acm_irq_cb_handler()`. 

https://github.com/zephyrproject-rtos/zephyr/blob/c710f8892beef3e486dded2f0ad4394e4a4d9aa0/subsys/usb/device_next/class/usbd_cdc_acm.c#L812C1-L813C32

The previous testing wasn't rigorous enough, that the subsequent transaction was actually triggered by new data coming in from RX instead (and hence the stale data was flushed out together). 

This being said, I have added another handling here to make sure that remaining data in tx_fifo, when the endpoint is unable to send it out in one shot, gets sent out immediately after current TX is completed as well. 

https://github.com/zephyrproject-rtos/zephyr/blob/a7aac8d37497e06ee64bf499edd9a38ce9dfdca6/subsys/usb/device_next/class/usbd_cdc_acm.c#L571C2-L580C3

Please help to review whether this make sense.

Reply:
I don't understand `cdc_acm_irq_cb_handler()`. Isn't the `data->tx_fifo.altered = false;` going to make the `if (data->tx_fifo.altered)` effectively a dead code? Shouldn't this be fixed instead of adding new boolean?

Reply:
> I don't understand `cdc_acm_irq_cb_handler()`. Isn't the `data->tx_fifo.altered = false;` going to make the `if (data->tx_fifo.altered)` effectively a dead code? Shouldn't this be fixed instead of adding new boolean?

I took me a while to understand this as well, from my understanding, in this handler, it will first clear all the altered flag, and then callback to the interrupt handler at application side: `data->cb(...)`. 

https://github.com/zephyrproject-rtos/zephyr/blob/a7aac8d37497e06ee64bf499edd9a38ce9dfdca6/subsys/usb/device_next/class/usbd_cdc_acm.c#L841-L844

If there is more data to be transmitted by the application, the `tx_fifo.altered` flag will be raised by application before the callback returns. Then here we go, continue to send data out again. 

I see that this part is checked in by @jfischer-no. @jfischer-no, can you please help to confirm whether this understanding is correct? 
Thanks.

Reply:
Isn't there a problem when there is more data in the tx ringbuffer? The sequence I think about is:
  * `cdc_acm_irq_cb_handler()` sets `data->tx_fifo.altered = false;`
  * `data->cb(usbd_class_get_private(c_data), data->cb_data);` is called, application writes to ring buffer more bytes than a CDC ACM TX buffer can fit
  * TX fifo work is triggered because `data->tx_fifo.altered` was set to true alongside ring buffer write
  * `cdc_acm_tx_fifo_handler()` calls `len = ring_buf_get(data->tx_fifo.rb, buf->data, buf->size);` and submits the transfer. The ring buffer still has pending TX data.
  * host reads all the data, `usbd_cdc_acm_request()` submits `cdc_acm_work_submit(&data->irq_cb_work);`
  * `cdc_acm_irq_cb_handler()` sets `data->tx_fifo.altered = false;`
  * `data->cb(usbd_class_get_private(c_data), data->cb_data);` is called, application doesn't have more data to write
  * TX fifo work is not triggered despite there being data in the ring buffer.

Is there anything preventing the above scenario?

Reply:
Hi @tmon-nordic, 

I added this for your case you mentioned in the latest commit just now ☺️ 

https://github.com/zephyrproject-rtos/zephyr/blob/a7aac8d37497e06ee64bf499edd9a38ce9dfdca6/subsys/usb/device_next/class/usbd_cdc_acm.c#L571-L580

Reply:
Why to have two booleans for roughly the same? Wouldn't it be better to change the `data->tx_fifo.altered` logic completely? If we go with separate variable, I think `bool tx_deferred;` should be moved inside `tx_fifo` structure.

But is it really necessary to have `tx_deferred`? I think that the `CDC_ACM_TX_EP_BUSY` in combination with `ring_buf_is_empty(data->tx_fifo.rb))` should be enough to determine whether the TX work has to be raised or not.

My idea would be to replace `if (data->tx_fifo.altered)` with `if (!ring_buf_is_empty(data->tx_fifo.rb)) && !atomic_test_and_set_bit(&data->state, CDC_ACM_TX_EP_BUSY))`. Then the `CDC_ACM_TX_EP_BUSY` would be cleared in `usbd_cdc_acm_request()` when the IN request completes. This would essentially eliminate the need for `data->tx_fifo.altered` altogether and potentially make the code simpler to follow and more robust. Am I missing something?

Reply:
Hi @tmon-nordic, I am new to open source development and zephyr OS as well. Wanted to check with you, is that okay to do such a change without consulting the original author?

Reply:
Open Source development works on technical merit. I would say that there is no need to consult the original author when proposing a change that solves issues or otherwise enhances the code base.

While there may be times where you want to ask what was intended, there is also the possibility that the original author simply is no longer around.

Reply:
sorry, was going to replace my comment with another one just now but somehow I lost all of them. 
(The comment on saying I am new to Open-Source and ZephyrOS, asking on how this community works)

I am kind of agree that in this case, checking for `ring_buf_is_empty()` at  `cdc_acm_irq_cb_handler()`  plus checking for `CDC_ACM_TX_EP_BUSY` flag at `cdc_acm_tx_fifo_handler()` would be optimal as checking the ring buffer status directly gives us the most accurate status for actions to be taken. 

It seems that the original intend of `altered` flag is to provide a fast access to the status whether application layer has accessed the ring buffer. However, it seems to me that the bit might not be good enough for the scenario in tx_fifo here anymore. 

@jfischer-no , what do you think? Is @tmon-nordic suggestion here to check whether to trigger `tx_fifo_work` based on whether `ring_buf_is_empty()` sounds good to you as well?

Reply:
Hi @tmon-nordic, changes has been made as discussed and tested working at my side. 
Please kindly review and see whether its all good for you. 
Thanks.